manifests/v1beta1/installs/katib-leader-election/katib-config.yaml

```diff
@@ -0,0 +1,62 @@
+# This KatibConfig is mostly same as https://github.com/kubeflow/katib/manifests/v1beta1/components/controller/katib-config.yaml.
+# Only `.init.controller.enableLeaderElection` field is different.
+---
```

Comment:
Unfortunately, Kustomize doesn't allow us to patch the ConfigMap data that is why we need to keep the whole config for each overlay 😞 
Which means we need to update the image tags for each release in the future in these overlays.
I think, we need to do something with this script: https://github.com/kubeflow/katib/blob/199aa8b963a85423309dd0d49b61b2cf5294079f/scripts/v1beta1/update-images.sh#L78-L81.

I think, at this stage it maybe worth to discuss how we want to maintain our suggestion image versions.
Currently, tags for suggestions we [maintain under `/manifests/v1beta1/components/controller/katib-config.yaml`](https://github.com/kubeflow/katib/blob/release-0.15/manifests/v1beta1/components/controller/katib-config.yaml#L11), tags for controller we [maintain under `/manifests/v1beta1/installs/katib-controller/kustomization.yaml`](https://github.com/kubeflow/katib/blob/release-0.15/manifests/v1beta1/installs/katib-standalone/kustomization.yaml#L25).

Reply:
What do you think @kubeflow/wg-automl-leads @tenzen-y ?

Reply:
> Which means we need to update the image tags for each release in the future in these overlays.
I think, we need to do something with this script:

It's a good point. 

> I think, at this stage it maybe worth to discuss how we want to maintain our suggestion image versions.
Currently, tags for suggestions we [maintain under /manifests/v1beta1/components/controller/katib-config.yaml](https://github.com/kubeflow/katib/blob/release-0.15/manifests/v1beta1/components/controller/katib-config.yaml#L11), tags for controller we [maintain under /manifests/v1beta1/installs/katib-controller/kustomization.yaml](https://github.com/kubeflow/katib/blob/release-0.15/manifests/v1beta1/installs/katib-standalone/kustomization.yaml#L25).

I think generating KatibConfig for each `installs` by script would be good. This means we generate `manifests/v1beta1/installs/katib-leader-election/katib-config.yaml` from `manifests/v1beta1/components/katib-config.yaml` by a scripts.

If so, we could maintain the suggestion service image versions the same way as in the past.

Reply:
> I think generating KatibConfig for each installs by script would be good. 

What do you mean by that ? 

I was thinking that since Kustomize doesn't allow us to patch the ConfgMap data, we can just replace it where it is required.
Similar what you did for [Katib Leader Election install.](https://github.com/kubeflow/katib/blob/199aa8b963a85423309dd0d49b61b2cf5294079f/manifests/v1beta1/installs/katib-leader-election/kustomization.yaml#L12-L18)
So for [Katib Standalone install](https://github.com/kubeflow/katib/blob/199aa8b963a85423309dd0d49b61b2cf5294079f/manifests/v1beta1/installs/katib-standalone/kustomization.yaml) and [Katib with Kubeflow install ](https://github.com/kubeflow/katib/blob/199aa8b963a85423309dd0d49b61b2cf5294079f/manifests/v1beta1/installs/katib-with-kubeflow/kustomization.yaml) we need to do the same to replace Katib Config with ConfigMap with appropriate image tags.
E.g. for `master` branch the image tags will be `:latest` for `release-0.x` branch the image tags will be `:v0.x.0`.

WDYT @tenzen-y @johnugeorge ?

Reply:
> I think generating KatibConfig for each installs by script would be good.
> 
> What do you mean by that ?

@andreyvelich NVM. My suggestion wouldn't work well.

> So for [Katib Standalone install](https://github.com/kubeflow/katib/blob/199aa8b963a85423309dd0d49b61b2cf5294079f/manifests/v1beta1/installs/katib-standalone/kustomization.yaml) and [Katib with Kubeflow install ](https://github.com/kubeflow/katib/blob/199aa8b963a85423309dd0d49b61b2cf5294079f/manifests/v1beta1/installs/katib-with-kubeflow/kustomization.yaml)we need to do the same to replace Katib Config with ConfigMap with appropriate image tags.

Does that mean that we use [kustomize imageTagTransformer](https://kubectl.docs.kubernetes.io/references/kustomize/builtins/#_imagetagtransformer_) like https://github.com/kubeflow/katib/blob/199aa8b963a85423309dd0d49b61b2cf5294079f/manifests/v1beta1/installs/katib-standalone/kustomization.yaml#L22-L34?

Kustomize recognizes `.data` field as string data, not structured data such as YAML. So we can not replace image tags on the `kind: KatibConfig` embedded on ConfigMap using imageTagTransformer.

So, we need to replace image tags using `vars` like https://github.com/kubeflow/katib/blob/199aa8b963a85423309dd0d49b61b2cf5294079f/manifests/v1beta1/installs/katib-with-kubeflow/kustomization.yaml#L46-L56.

Reply:
The kustomize vars solution: https://github.com/kubeflow/katib/pull/2176/commits/19ebc22ff2600aa7a83dc99ce9b58a97adf4cc35.

Reply:
@tenzen-y I meant, that for every install we are going to replace the whole Katib Config, similar to [Katib Leader Election](https://github.com/kubeflow/katib/blob/6c76383507a082ba91edcfaab32c73c669b304a3/manifests/v1beta1/installs/katib-leader-election/kustomization.yaml#L12-L18). E.g. for Katib Standalone we just have the same config as in `/components/katib-config`, but it will have different image tags (for `master` branch latest tag, for `release-` branch `v0.x` tag. Does it make sense? 

I like the idea with `imageTagTransformers`, but would it be confusing for our users ? Especially for those who are not familiar with Kustomize.
I think, the idea of installs/overlays was to have 1 place where we set settings, images, or any other changes for our Katib Control Plane.

Other solution could be to just remove Katib Config from `components/` manifests and add it only in `installs`. [We can modify our docs to explain it.](https://www.kubeflow.org/docs/components/katib/hyperparameter/#installing-katib)

What do you think @johnugeorge @tenzen-y ?

Reply:
> I meant, that for every install we are going to replace the whole Katib Config, similar to [Katib Leader Election](https://github.com/kubeflow/katib/blob/6c76383507a082ba91edcfaab32c73c669b304a3/manifests/v1beta1/installs/katib-leader-election/kustomization.yaml#L12-L18). E.g. for Katib Standalone we just have the same config as in /components/katib-config, but it will have different image tags (for master branch latest tag, for release- branch v0.x tag. Does it make sense?

@andreyvelich Oh, I see. Thanks for the clarification.

> I think, the idea of installs/overlays was to have 1 place where we set settings, images, or any other changes for our Katib Control Plane.
> 
> Other solution could be to just remove Katib Config from components/ manifests and add it only in installs. [We can modify our docs to explain it.](https://www.kubeflow.org/docs/components/katib/hyperparameter/#installing-katib)

It sounds reasonable. I will do it ASAP.

Reply:
@johnugeorge @gaocegege I am happy to discuss any other suggestions how to configure our Katib manifests.

Reply:
@andreyvelich I've done it: https://github.com/kubeflow/katib/pull/2176/commits/42fe2783ed7124f7a3c6717eb902af0a1ef7abec
Is this your expected?

Reply:
Yeah, that's right @tenzen-y, that what was I proposed.

Reply:
Thanks for the check :)